File: class17/java/PES1701-Locadora.java
/*
 * PES1701 - Mini-Sistema Locadora de Veículos
 * Aula 17: Projeto Parcial — Modelagem UML + Implementação OO
 * Projeto e Engenharia de Software - Prof. Giovane Barcelos
 *
 * Sistema completo demonstrando herança, polimorfismo e encapsulamento
 * em um domínio de locadora de veículos (Carro, Moto).
 *
 * Compilar/executar:
 *   javac PES1701-Locadora.java && java Locadora
 */

import java.util.*;

abstract class Veiculo {
    private String placa, modelo;
    private double valorDiaria;

    Veiculo(String placa, String modelo, double valorDiaria) {
        this.placa = placa;
        this.modelo = modelo;
        this.valorDiaria = valorDiaria;
    }

    public String getPlaca() { return placa; }
    public String getModelo() { return modelo; }
    public double getValorDiaria() { return valorDiaria; }

    public abstract double calcularLocacao(int dias);

    @Override
    public String toString() {
        return String.format("%s[%s] %s", getClass().getSimpleName(), placa, modelo);
    }
}

class Carro extends Veiculo {
    private int numPortas;

    Carro(String placa, String modelo, double valorDiaria, int numPortas) {
        super(placa, modelo, valorDiaria);
        this.numPortas = numPortas;
    }

    @Override
    public double calcularLocacao(int dias) {
        return getValorDiaria() * dias * 1.10;
    }
}

class Moto extends Veiculo {
    private int cilindradas;

    Moto(String placa, String modelo, double valorDiaria, int cilindradas) {
        super(placa, modelo, valorDiaria);
        this.cilindradas = cilindradas;
    }

    @Override
    public double calcularLocacao(int dias) {
        return getValorDiaria() * dias;
    }
}

class Locacao {
    private Veiculo veiculo;
    private String cliente;
    private int dias;
    private double valorTotal;

    Locacao(Veiculo veiculo, String cliente, int dias) {
        this.veiculo = veiculo;
        this.cliente = cliente;
        this.dias = dias;
        this.valorTotal = veiculo.calcularLocacao(dias);
    }

    double getValorTotal() { return valorTotal; }

    void resumo() {
        System.out.printf("  Veículo: %s%n", veiculo);
        System.out.printf("  Cliente: %s%n", cliente);
        System.out.printf("  Dias: %d%n", dias);
        System.out.printf("  Valor Total: R$ %.2f%n", valorTotal);
    }
}

class Locadora {
    private List<Veiculo> frota = new ArrayList<>();
    private List<Locacao> locacoes = new ArrayList<>();

    void cadastrar(Veiculo veiculo) {
        frota.add(veiculo);
        System.out.printf("  ✓ Cadastrado: %s%n", veiculo);
    }

    void listarDisponiveis() {
        System.out.printf("%n  --- Frota (%d veículos) ---%n", frota.size());
        for (Veiculo v : frota) {
            System.out.printf("  • %s — R$ %.2f/dia%n", v, v.getValorDiaria());
        }
    }

    Locacao alugar(String placa, String cliente, int dias) {
        for (Veiculo v : frota) {
            if (v.getPlaca().equals(placa)) {
                Locacao loc = new Locacao(v, cliente, dias);
                locacoes.add(loc);
                System.out.printf("%n  ✓ Locação registrada:%n");
                loc.resumo();
                return loc;
            }
        }
        System.out.printf("  ✗ Veículo %s não encontrado.%n", placa);
        return null;
    }

    double faturamentoTotal() {
        return locacoes.stream().mapToDouble(Locacao::getValorTotal).sum();
    }
}

class LocadoraDemo {
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("  LOCADORA DE VEÍCULOS — SISTEMA OO");
        System.out.println("=".repeat(60));

        Locadora locadora = new Locadora();

        System.out.printf("%n--- Cadastro de Veículos ---%n");
        locadora.cadastrar(new Carro("ABC-1234", "Fiat Uno", 120.00, 4));
        locadora.cadastrar(new Carro("DEF-5678", "Honda Civic", 200.00, 4));
        locadora.cadastrar(new Moto("XYZ-9012", "Honda CG 160", 80.00, 160));

        locadora.listarDisponiveis();

        System.out.printf("%n--- Locações ---%n");
        locadora.alugar("ABC-1234", "Ana Silva", 5);
        locadora.alugar("XYZ-9012", "Carlos Santos", 3);

        System.out.printf("%n  ✓ Faturamento Total: R$ %.2f%n", locadora.faturamentoTotal());
        System.out.printf("%n✓ Sistema completo com herança, polimorfismo e encapsulamento.%n");
    }
}

File: class17/python/PES1701-Locadora.py
"""
PES1701 - Mini-Sistema Locadora de Veículos
Aula 17: Projeto Parcial — Modelagem UML + Implementação OO
Projeto e Engenharia de Software - Prof. Giovane Barcelos

Sistema completo demonstrando herança, polimorfismo e encapsulamento
em um domínio de locadora de veículos (Carro, Moto).
"""

from abc import ABC, abstractmethod


class Veiculo(ABC):
    """Classe abstrata base para todos os veículos."""

    def __init__(self, placa, modelo, valor_diaria):
        self.__placa = placa
        self.__modelo = modelo
        self.__valor_diaria = float(valor_diaria)

    @property
    def placa(self):
        return self.__placa

    @property
    def modelo(self):
        return self.__modelo

    @property
    def valor_diaria(self):
        return self.__valor_diaria

    @abstractmethod
    def calcular_locacao(self, dias):
        """Polimorfismo: cada tipo de veículo tem sua regra de cálculo."""
        pass

    def __repr__(self):
        return f"{self.__class__.__name__}[{self.__placa}] {self.__modelo}"


class Carro(Veiculo):
    """Carro: valor = diária × dias + taxa de seguro (10%)."""

    def __init__(self, placa, modelo, valor_diaria, num_portas=4):
        super().__init__(placa, modelo, valor_diaria)
        self.num_portas = num_portas

    def calcular_locacao(self, dias):
        return self.valor_diaria * dias * 1.10  # +10% seguro


class Moto(Veiculo):
    """Moto: valor = diária × dias (sem seguro adicional)."""

    def __init__(self, placa, modelo, valor_diaria, cilindradas=150):
        super().__init__(placa, modelo, valor_diaria)
        self.cilindradas = cilindradas

    def calcular_locacao(self, dias):
        return self.valor_diaria * dias


class Locacao:
    """Registra uma locação de veículo."""

    def __init__(self, veiculo, cliente, dias):
        self.__veiculo = veiculo
        self.__cliente = cliente
        self.__dias = dias
        self.__valor_total = veiculo.calcular_locacao(dias)

    @property
    def valor_total(self):
        return self.__valor_total

    def resumo(self):
        print(f"  Veículo: {self.__veiculo}")
        print(f"  Cliente: {self.__cliente}")
        print(f"  Dias: {self.__dias}")
        print(f"  Valor Total: R$ {self.__valor_total:.2f}")


class Locadora:
    """Gerencia a frota e as locações."""

    def __init__(self):
        self.__frota = []
        self.__locacoes = []

    def cadastrar(self, veiculo):
        self.__frota.append(veiculo)
        print(f"  ✓ Cadastrado: {veiculo}")

    def listar_disponiveis(self):
        print(f"\n  --- Frota ({len(self.__frota)} veículos) ---")
        for v in self.__frota:
            print(f"  • {v} — R$ {v.valor_diaria:.2f}/dia")

    def alugar(self, placa, cliente, dias):
        for v in self.__frota:
            if v.placa == placa:
                loc = Locacao(v, cliente, dias)
                self.__locacoes.append(loc)
                print(f"\n  ✓ Locação registrada:")
                loc.resumo()
                return loc
        print(f"  ✗ Veículo {placa} não encontrado.")
        return None

    def faturamento_total(self):
        return sum(loc.valor_total for loc in self.__locacoes)


if __name__ == "__main__":
    print("=" * 60)
    print("  LOCADORA DE VEÍCULOS — SISTEMA OO")
    print("=" * 60)

    locadora = Locadora()

    # Cadastro
    print("\n--- Cadastro de Veículos ---")
    locadora.cadastrar(Carro("ABC-1234", "Fiat Uno", 120.00, 4))
    locadora.cadastrar(Carro("DEF-5678", "Honda Civic", 200.00, 4))
    locadora.cadastrar(Moto("XYZ-9012", "Honda CG 160", 80.00, 160))

    locadora.listar_disponiveis()

    # Locações
    print("\n--- Locações ---")
    locadora.alugar("ABC-1234", "Ana Silva", 5)
    locadora.alugar("XYZ-9012", "Carlos Santos", 3)

    print(f"\n  ✓ Faturamento Total: R$ {locadora.faturamento_total():.2f}")
    print("\n✓ Sistema completo com herança, polimorfismo e encapsulamento.")
